File: brique-102/packages/python/molam/resources/webhooks.py
"""
Webhooks resource
"""

import hmac
import hashlib
import time
from typing import Dict, Any, List


class Webhooks:
    """Webhooks API"""

    def __init__(self, http):
        self.http = http

    def verify_signature(
        self,
        raw_body: str,
        signature_header: str,
        secret: str
    ) -> bool:
        """
        Verify webhook signature

        Args:
            raw_body: Raw request body as string
            signature_header: Molam-Signature header value
            secret: Webhook secret

        Returns:
            True if signature is valid

        Raises:
            ValueError: If signature is invalid or expired
        """
        parts = dict(p.split("=", 1) for p in signature_header.split(","))

        timestamp_str = parts.get("t")
        signature = parts.get("v1")

        if not timestamp_str or not signature:
            raise ValueError("Invalid signature header format")

        timestamp = int(timestamp_str)
        now = int(time.time() * 1000)

        # Check timestamp (5-minute tolerance)
        if abs(now - timestamp) > 5 * 60 * 1000:
            raise ValueError("Signature timestamp outside tolerance")

        # Compute HMAC
        payload = f"{timestamp_str}.{raw_body}"
        computed = hmac.new(
            secret.encode(),
            payload.encode(),
            hashlib.sha256
        ).hexdigest()

        # Constant-time comparison
        if not hmac.compare_digest(computed, signature):
            raise ValueError("Signature mismatch")

        return True

    def create_endpoint(
        self,
        tenant_type: str,
        tenant_id: str,
        url: str,
        events: List[str]
    ) -> Dict[str, Any]:
        """
        Create webhook endpoint

        Args:
            tenant_type: Tenant type
            tenant_id: Tenant ID
            url: Webhook URL
            events: List of event types to subscribe to

        Returns:
            Endpoint object
        """
        payload = {
            "tenant_type": tenant_type,
            "tenant_id": tenant_id,
            "url": url,
            "events": events
        }
        return self.http.post("/v1/webhooks/endpoints", payload)

    def list_endpoints(self, tenant_type: str, tenant_id: str) -> Dict[str, Any]:
        """
        List webhook endpoints

        Args:
            tenant_type: Tenant type
            tenant_id: Tenant ID

        Returns:
            List of endpoints
        """
        return self.http.get(
            f"/v1/webhooks/endpoints?tenant_type={tenant_type}&tenant_id={tenant_id}"
        )

    def delete_endpoint(self, endpoint_id: str) -> Dict[str, Any]:
        """
        Delete webhook endpoint

        Args:
            endpoint_id: Endpoint ID

        Returns:
            Deletion confirmation
        """
        return self.http.delete(f"/v1/webhooks/endpoints/{endpoint_id}")

File: brique-102/packages/php/src/Resources/Webhooks.php
<?php

namespace Molam\Resources;

use Molam\Http\HttpClient;

class Webhooks
{
    /** @var HttpClient */
    private $http;

    public function __construct(HttpClient $http)
    {
        $this->http = $http;
    }

    /**
     * Verify webhook signature
     *
     * @param string $rawBody Raw request body
     * @param string $signatureHeader Molam-Signature header value
     * @param string $secret Webhook secret
     * @return bool
     * @throws \Exception
     */
    public function verifySignature(string $rawBody, string $signatureHeader, string $secret): bool
    {
        $parts = [];
        foreach (explode(',', $signatureHeader) as $part) {
            [$key, $value] = explode('=', $part, 2);
            $parts[$key] = $value;
        }

        if (!isset($parts['t'], $parts['v1'])) {
            throw new \Exception('Invalid signature header format');
        }

        // Check timestamp (5-minute tolerance)
        $timestamp = (int)$parts['t'];
        $now = (int)(microtime(true) * 1000);

        if (abs($now - $timestamp) > 5 * 60 * 1000) {
            throw new \Exception('Signature timestamp outside tolerance');
        }

        // Compute HMAC
        $computed = hash_hmac('sha256', $parts['t'] . '.' . $rawBody, $secret);

        // Constant-time comparison
        if (!hash_equals($computed, $parts['v1'])) {
            throw new \Exception('Signature mismatch');
        }

        return true;
    }

    /**
     * Create webhook endpoint
     */
    public function createEndpoint(string $tenantType, string $tenantId, string $url, array $events): array
    {
        $payload = [
            'tenant_type' => $tenantType,
            'tenant_id' => $tenantId,
            'url' => $url,
            'events' => $events
        ];

        return $this->http->post('/v1/webhooks/endpoints', $payload);
    }

    /**
     * List webhook endpoints
     */
    public function listEndpoints(string $tenantType, string $tenantId): array
    {
        return $this->http->get("/v1/webhooks/endpoints?tenant_type={$tenantType}&tenant_id={$tenantId}");
    }

    /**
     * Delete webhook endpoint
     */
    public function deleteEndpoint(string $endpointId): array
    {
        return $this->http->delete("/v1/webhooks/endpoints/{$endpointId}");
    }
}
